src/cel3ds/3dsread.cpp

```diff
@@ -34,35 +35,39 @@ using ProcessChunkFunc = std::int32_t (*)(std::istream &, std::uint16_t, std::in
 
 bool readInt(std::istream& in, std::int32_t& value)
 {
-    in.read(reinterpret_cast<char*>(&value), sizeof(std::int32_t));
-    if (!in.good()) { return false; }
+    char buffer[sizeof(value)];
```

Comment:
Could you point to any official documentation regarding UB in read()? Because https://en.cppreference.com/w/cpp/io/basic_istream/read uses exactly the same code as the current Celestia.

Reply:
It's not precisely the same, but this appears to be similar to the example at 43:44 in Timur Doumler's talk "[Type Punning in Modern C++](https://www.youtube.com/watch?v=_qzMpk-22cc)" - while the `reinterpret_cast` from a `std::int32_t*` to `char*` is allowed, according to the C++ abstract machine the resultant pointer does not actually point to a char array (and there isn't a char array at this location), therefore any access to it as though it was a char array is (as far as I understand it) undefined behaviour.

Reply:
```
    std::uint32_t n;
    if(raw.read(reinterpret_cast<char*>(&n), sizeof n))
        std::cout << std::hex << std::showbase << n << '\n';
```

is this not the same?

i need a few days to check the video, unfortunately i don't have time till Monday

Reply:
As far as I'm aware, that example is undefined behaviour because the first parameter of `read` should be a pointer to a char array, and by the rules of the C++ abstract machine, there isn't actually a char array at whatever `reinterpret_cast<char*>(&n)` is pointing to.

Reply:
@ajtribick can you remove this commit from this PR so we can go forward?

Reply:
I've just found what i was looking for: https://en.cppreference.com/w/c/language/object, "Strict aliasing".
> Given an object with effective type T1, using an lvalue expression (typically, dereferencing a pointer) of a different type T2 is undefined behavior, unless:
> * T2 is a character type (char, signed char, or unsigned char).

So our old code is absolutely correct.

Reply:
I've just found what i was looking for: https://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing.

> Whenever an attempt is made to read or modify the stored value of an object of type DynamicType through a glvalue of type AliasedType, the behavior is undefined unless one of the following is true: 
> AliasedType is std::byte (since C++17), char, or unsigned char: this permits examination of the object representation of any object as an array of bytes.

So our old code is absolutely correct.

Reply:
I'll point out that cppreference is not the standard. I fully agree that `reinterpret_cast` rules regarding `char` and `unsigned char` are *intended* to let you access the byte representation of the array but due to rules brought in in C++17 this seems to have accidentally become undefined behaviour. The proposal to fix this is http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1839r2.pdf which includes a description of the problem.

From what I'm seeing in that proposal, I *think* we're ok with `reinterpret_cast` in C++14, but I think it would be good to stay forward-compatible with the C++17 standard where the only valid way of accessing the byte representation of an object is via `memcpy`. Based on what I'm seeing on Godbolt the optimisers in the major compilers are pretty good at removing the `memcpy` from the final code.